packages/next/src/build/handle-externals.ts

```diff
@@ -374,28 +359,86 @@ export function makeExternalHandler({
       }
     }
 
-    const shouldBeBundled =
-      isResourceInPackages(
-        res,
-        config.transpilePackages,
-        resolvedExternalPackageDirs
-      ) ||
-      (isEsm && isAppLayer) ||
-      (!isAppLayer && config.experimental.bundlePagesExternals)
-
-    if (nodeModulesRegex.test(res)) {
-      if (isWebpackServerLayer(layer)) {
-        if (!optOutBundlingPackageRegex.test(res)) {
-          return // Bundle for server layer
-        }
-        return `${externalType} ${request}` // Externalize if opted out
-      }
+    const resolvedBundlingOptOutRes = resolveBundlingOptOutPackages({
+      resolvedRes: res,
+      optOutBundlingPackageRegex,
+      config,
+      resolvedExternalPackageDirs,
+      isEsm,
+      isAppLayer,
+      layer,
+      externalType,
+      request,
+    })
+    if (resolvedBundlingOptOutRes) {
+      return resolvedBundlingOptOutRes
+    }
+
+    // if here, we default to bundling the file
+    return
+  }
+}
 
-      if (!shouldBeBundled || optOutBundlingPackageRegex.test(res)) {
-        return `${externalType} ${request}` // Externalize if not bundled or opted out
+function resolveBundlingOptOutPackages({
+  resolvedRes,
+  optOutBundlingPackageRegex,
+  config,
+  resolvedExternalPackageDirs,
+  isEsm,
+  isAppLayer,
+  layer,
+  externalType,
+  request,
+}: {
+  resolvedRes: string
+  optOutBundlingPackageRegex: RegExp
+  config: NextConfigComplete
+  resolvedExternalPackageDirs: Map<string, string>
+  isEsm: boolean
+  isAppLayer: boolean
+  layer: WebpackLayerName | null
+  externalType: string
+  request: string
+}) {
+  const shouldBeBundled =
+    isResourceInPackages(
+      resolvedRes,
+      config.transpilePackages,
+      resolvedExternalPackageDirs
+    ) ||
+    (isEsm && isAppLayer) ||
+    (!isAppLayer && config.experimental.bundlePagesExternals)
+
+  const isOptOutBundling = optOutBundlingPackageRegex.test(resolvedRes)
```

Comment:
Might be able to move this into the `if (nodeModulesRegex.test(resolvedRes)) {` below so no unnecessary regex executions.